### PR TITLE
Fixing categories

### PR DESCRIPTION
**Pull request by KarimChukfeh**
I wrote couple migrations to make every Category belongs_to CategoryOption. Further I added a CategoryOption.repositories method. Next I'll add a migration to get of category column in Repository and make these old records use the new Category system.

### CODE REVIEW

File: app/controllers/repositories_controller.rb [hunk 1]
```diff
@@ -65,7 +65,7 @@ def create
       create_categories
       create_equipments
       render json: { redirect_uri: "#{repository_path(@user.username, @repository.slug)}" }
-      Repository.reindex
+      # Repository.reindex
```
Comment (tahnok) [inline app/controllers/repositories_controller.rb]: probably worth deleting instead of just commenting. I think this was a solr thing

File: app/models/category_option.rb [hunk 2]
```diff
@@ -1,5 +1,10 @@
 class CategoryOption < ActiveRecord::Base
   belongs_to :admin
-  validates :name, presence: true
+  validates :name, presence: true, uniqueness: true
   scope :show_options, -> { order("lower(name) ASC").all }
+
+  def repositories
```
Comment (tahnok) [inline app/models/category_option.rb]: I think you can replace this with `has_many :categories`, `has_many :repositories, through: categories`

File: db/migrate/20170706152142_add_category_option_id_to_categories.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+class AddCategoryOptionIdToCategories < ActiveRecord::Migration
+  def change
+    Category.all.each do |cat|
+      cat.category_option_id = CategoryOption.find_by(name: cat.name).id
+      cat.save
```
Comment (tahnok) [inline db/migrate/20170706152142_add_category_option_id_to_categories.rb]: It would be safer to do `save!` which will raise an exception and stop if something happens

File: db/schema.rb [hunk 4]
```diff
@@ -33,9 +34,9 @@
 
   create_table "certifications", force: :cascade do |t|
     t.integer  "user_id"
-    t.string   "name"
-    t.datetime "created_at", null: false
-    t.datetime "updated_at", null: false
+    t.datetime "created_at",          null: false
+    t.datetime "updated_at",          null: false
+    t.integer  "training_session_id"
```
Comment (tahnok) [inline db/schema.rb]: You may have to reset your database and then run `rake db:schema:dump` because this shouldn't be here

File: test/controllers/repositories_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
+require 'test_helper'
+
+class RepositoriesControllerTest < ActionController::TestCase
+
+  test "can create new repository" do
```
Comment (tahnok) [inline test/controllers/repositories_controller_test.rb]: :+1:

File: app/controllers/repositories_controller.rb [hunk 6]
```diff
@@ -174,9 +172,12 @@ def update_files
     end
 
     def create_categories
-      params['categories'].first(5).each do |c|
-        Category.create(name: c, repository_id: @repository.id)
-      end if params['categories'].present?
+      if params['categories'].present?
+        cats = params['categories']
+        cats.each do |cat|
+          Category.create(category_option_id: CategoryOption.find_by(name: String(cat)).id, repository_id: @repository.id)
```
Comment (tahnok) [inline app/controllers/repositories_controller.rb]: Why are you casting cat to a string? The ruby way to do this is `cat.to_s`

File: db/migrate/20170706152142_add_category_option_id_to_categories.rb [hunk 7]
```diff
@@ -0,0 +1,8 @@
+class AddCategoryOptionIdToCategories < ActiveRecord::Migration
+  def change
+    Category.all.each do |cat|
+      cat.category_option_id = CategoryOption.find_by(name: cat.name).id
```
Comment (tahnok) [inline db/migrate/20170706152142_add_category_option_id_to_categories.rb]: I think this might not work properly with the addition of the `name` method to category. Without the manual addition of `name`, `cat.name` will fetch the field in the database. After adding the new `name` method `cat.name` will fetch the category_option and get it's name which will be null here.

I think the solution is to add `name` to category in another PR after this one has been merged and deployed

File: db/migrate/20170706152142_add_category_option_id_to_categories.rb [hunk 8]
```diff
@@ -0,0 +1,8 @@
+class AddCategoryOptionIdToCategories < ActiveRecord::Migration
+  def change
```
Comment (tahnok) [inline db/migrate/20170706152142_add_category_option_id_to_categories.rb]: You should change this from `def change` to `def up` and `def down` like in the examples section of http://api.rubyonrails.org/classes/ActiveRecord/Migration.html

File: db/migrate/20170706172353_remove_category_column_from_repositories.rb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+class RemoveCategoryColumnFromRepositories < ActiveRecord::Migration
+  def change
+    remove_column :repositories, :category, :string
```
Comment (tahnok) [inline db/migrate/20170706172353_remove_category_column_from_repositories.rb]: This needs to be in a separate PR since you have code in here that depends on this

File: app/controllers/repositories_controller.rb (hunk 6 above)
Comment (KarimChukfeh) [inline app/controllers/repositories_controller.rb]: it didn't work before casting to string but now it does 👍

File: db/migrate/20170706152142_add_category_option_id_to_categories.rb (hunk 8 above)
Comment (KarimChukfeh) [inline db/migrate/20170706152142_add_category_option_id_to_categories.rb]: done

File: db/migrate/20170706152142_add_category_option_id_to_categories.rb (hunk 7 above)
Comment (KarimChukfeh) [inline db/migrate/20170706152142_add_category_option_id_to_categories.rb]: Sounds good

File: db/migrate/20170706151010_add_category_option_to_categories.rb [hunk 10]
```diff
@@ -1,5 +1,5 @@
 class AddCategoryOptionToCategories < ActiveRecord::Migration
-  def change
+  def up
```
Comment (tahnok) [inline db/migrate/20170706151010_add_category_option_to_categories.rb]: I think you need to have a `down` method too, but it can be empty

File: db/migrate/20170706172353_remove_category_column_from_repositories.rb [hunk 11]
```diff
@@ -0,0 +1,5 @@
+class RemoveCategoryColumnFromRepositories < ActiveRecord::Migration
+  def down
```
Comment (tahnok) [inline db/migrate/20170706172353_remove_category_column_from_repositories.rb]: This should be part of 'up' since it will happen when the migration is 'brought up'. Right now it will only run if you `rake db:rollback` once this has been migrated

**Comment by tahnok** [comment]
Is there something I can do to help you break this down into smaller pieces? Once that happens we can start merging each piece and deploying one by one

**Comment by tahnok** [comment]
Oh, I'm an idiot and didn't see some of your other PRs. NEVERMIND!
